src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -107,32 +117,278 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
                     return null;
                 }
 
-                // Walk back through the chain of ConcatNIterators looking for the one
+                // Walk back through the chain of ConcatNEnumerableIterators looking for the one
                 // that has its _nextIndex equal to index.  If we don't find one, then it
                 // must be prior to any of them, so call GetEnumerable on the previous
-                // Concat2Iterator.  This avoids a deep recursive call chain.
-                ConcatNIterator<TSource> current = this;
+                // Concat2EnumerableIterator.  This avoids a deep recursive call chain.
+                ConcatNEnumerableIterator<TSource> current = this;
                 while (true)
                 {
                     if (index == current._nextIndex)
                     {
                         return current._next;
                     }
 
-                    ConcatNIterator<TSource> prevN = current._previousConcat as ConcatNIterator<TSource>;
+                    var prevN = current._previousConcat as ConcatNEnumerableIterator<TSource>;
                     if (prevN != null)
                     {
                         current = prevN;
                         continue;
                     }
 
-                    Debug.Assert(current._previousConcat is Concat2Iterator<TSource>);
+                    ConcatIterator<TSource> prev2 = current._previousConcat;
+                    Debug.Assert(prev2 is Concat2EnumerableIterator<TSource> || prev2 is Concat2CollectionIterator<TSource>);
                     Debug.Assert(index == 0 || index == 1);
-                    return current._previousConcat.GetEnumerable(index);
+                    return prev2.GetEnumerable(index);
                 }
             }
         }
 
+        private sealed class Concat2CollectionIterator<TSource> : ConcatCollectionIterator<TSource>
+        {
+            private readonly ICollection<TSource> _first;
+            private readonly ICollection<TSource> _second;
+
+            internal Concat2CollectionIterator(ICollection<TSource> first, ICollection<TSource> second)
+            {
+                Debug.Assert(first != null && second != null);
+                _first = first;
+                _second = second;
+            }
+
+            internal override int Count => checked(_first.Count + _second.Count);
```

Comment:
Does Concat(collection1, collection2).Count() throw on overflow today?  Please add a test for it and make sure the behavior stays the same before and after your change.